File: sdks/go/features/ratelimits.go
// package features provides functionality for interacting with hatchet features.
package features

import (
	"context"

	"github.com/cockroachdb/errors"
	"github.com/google/uuid"

	v0Client "github.com/hatchet-dev/hatchet/pkg/client"
	"github.com/hatchet-dev/hatchet/pkg/client/rest"
	"github.com/hatchet-dev/hatchet/pkg/client/types"
)

// CreateRatelimitOpts contains options for creating or updating a rate limit.
type CreateRatelimitOpts struct {
	Key      string
	Duration types.RateLimitDuration
	Limit    int
}

// RateLimitsClient provides methods for interacting with rate limits
type RateLimitsClient struct {
	api      *rest.ClientWithResponses
	admin    v0Client.AdminClient
	tenantId uuid.UUID
}

// NewRateLimitsClient creates a new RateLimitsClient with the provided api client, tenant id, and admin client.
func NewRateLimitsClient(
	api *rest.ClientWithResponses,
	tenantId string,
	admin v0Client.AdminClient,
) *RateLimitsClient {
	tenantIdUUid := uuid.MustParse(tenantId)

	return &RateLimitsClient{
		api:      api,
		tenantId: tenantIdUUid,
		admin:    admin,
	}
}

// Upsert creates or updates a rate limit with the provided options.
func (c *RateLimitsClient) Upsert(opts CreateRatelimitOpts) error {
	if err := c.admin.PutRateLimit(opts.Key, &types.RateLimitOpts{
		Max:      opts.Limit,
		Duration: opts.Duration,
	}); err != nil {
		return errors.Wrap(err, "failed to upsert rate limit")
	}

	return nil
}

// List retrieves rate limits based on the provided parameters (optional).
func (c *RateLimitsClient) List(ctx context.Context, opts *rest.RateLimitListParams) (*rest.RateLimitList, error) {
	resp, err := c.api.RateLimitListWithResponse(
		ctx,
		c.tenantId,
		opts,
	)
	if err != nil {
		return nil, errors.Wrap(err, "failed to list rate limits")
	}

	if err := validateJSON200Response(resp.StatusCode(), resp.Body, resp.JSON200); err != nil {
		return nil, err
	}

	return resp.JSON200, nil
}

File: sdks/typescript/src/core/features/rate-limits.ts
import { retrier, type RetrierConfig } from '@hatchet/util/retrier';
import type { Logger } from '@hatchet/util/logger/logger';
import type { WorkflowServiceClient } from '@hatchet/protoc/workflows/workflows';
import type { CreateRateLimitOpts } from '../types';

export interface RateLimitsClientDeps {
  rpc: WorkflowServiceClient;
  logger: Logger;
  retrier?: RetrierConfig;
}

/** Tenant-wide rate limits that tasks declare against by key. */
export class RateLimitsClient {
  constructor(private readonly deps: RateLimitsClientDeps) {}

  /**
   * Creates or updates a rate limit.
   * @returns the rate limit's key
   */
  async put(opts: CreateRateLimitOpts): Promise<string> {
    await retrier(
      () =>
        this.deps.rpc.putRateLimit({ key: opts.key, limit: opts.limit, duration: opts.duration }),
      this.deps.logger,
      this.deps.retrier
    );
    return opts.key;
  }
}
